### PR TITLE
Sub domain allocation plan

### PR DESCRIPTION
**Pull request by betatim**
This PR makes a start on capturing the ideas from #144 and writing them down in our documentation.

I tried to formulate them as "you should do this" instead of rules or negatives.

What do you think?

### CODE REVIEW

File: docs/binder/subdomains.rst [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+mention of who is operating this hub and/or sponsoring the resources is welcome.
+The hub should not include adverts, cryptominers or other malware in pages it
+serves.
+
+The operator has to be committed to and able to operate a public BinderHub for a
+significant amount of time (at least a few months).
+
+The mission of the operator should be aligned with the mission of the Binder
+Project laid out in :ref:`binder-governance`. By allocating a sub-domain the
+project wants to recognise those who publicly support the project inn order
```
Comment (consideRatio) [inline docs/binder/subdomains.rst]:
```suggestion
project wants to recognise those who publicly support the project in order
```

Comment (betatim) [reply docs/binder/subdomains.rst]: Thanks! I need to investigate how to fix up my n key, it double types. Guess what brand of laptop I have ...

**Comment by choldgraf** [comment]
The only thing I'd add is something to describe what happens if any of those conditions are *not* met (e.g. "The mybinder.org team reserves the right to remove any sub-domains (or redirect them to mybinder.org) in the event that the BinderHub deployment behind a sub-domain is not following these guidelines to an acceptable level" or something like that. What do you think?

**Comment by betatim** [comment]
Not sure. For me it is implied that if you "fall out of compliance and fail to rectify the situation" you are out. If we start spelling out the rules too much do we encourage rule lawyering? I think that is what I'd worry about. Leaving it somewhat vague/not read like a legal text makes me hope that people will read this in the spirit it is intended and not with a thought towards "how can we exploit the rules".

In the end we control mybinder.org so we can do as we please, the worst outcome is that someone is annoyed that we redirected blah.mybinder.org back to mybinder.org. If they are the kind of person to get annoyed at that they'll probably get annoyed even if there are rules because they'll argue they were following them and ... 🤷‍♂️

If you propose a concrete sentence I'll definitely add it, otherwise I'll wait and see if someone else has an opinion either way.

**Comment by minrk** [comment]
Something like:

> If the BinderHub stops being functional, or is deemed to not behave in an acceptable way, the subdomain may be reclaimed at any time, redirecting back to the root mybinder.org.

**Comment by betatim** [comment]
Added Min's sentence.

**Comment by arnim** [comment]
> The BinderHub should maintain the look and feel of a stock BinderHub"

Would we (or Pangeo) satisfy this condition?

**Comment by betatim** [comment]
> Would we (or Pangeo) satisfy this condition?

The short answer is: I don't know.

I think we should (edit and then) merge or reject this, then field applications and discuss those in a separate thread.

My feeling is that we will encounter lots of things we didn't think of when we try to use this process. Instead of trying to cover them all up front I'd be Ok with waiting till we encounter them and then amend this document (when needed).

(For example I just wondered what should happen if a hub is at xxx.mybinder.org and users generate binder badges on it. Should they point to xxx.mybinder.org or mybinder.org? What about hubs hosted at a URL prefix that isn't just `/`?)

**Comment by arnim** [comment]
>  what should happen if a hub is at xxx.mybinder.org and users generate binder badges on it. Should they point to xxx.mybinder.org or mybinder.org? 

@betatim I would like if we all would just use mybinder.org . This would not only help to prevent fracturing of the ecosystem but also enable things like load balancing (e.g. depending on the usage/health of the different instances) in the future.  
However, for that to be possible I think it would be necessary for the new instances (sub domain users) not to be easily banned. If an institution wants to build its service offering (or part of it) around reproducible research with a binder, having an outside entity that has a kill switch would potentially make it difficult for these institutions to invest in the system.

However, I don't have a simple answer there either.

**Comment by choldgraf** [comment]
Two thoughts:

1. I agree with @betatim that this PR in particular is nearly ready to go. I think we should keep an eye on how things go and plan to revisit this language sometime in the future when we have more experience.

2. I agree with @arnim that we should keep binder links attached to the parent URL as much as possible. I remember @Carreau had a POC implementation to re-direct based on cookies, maybe something like that can be used (e.g. you have a "preferred binder" list)

File: docs/binder/subdomains.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+- making the BinderHub it points to easier to discover;
+- letting the user know which BinderHub they are being served by; and
+- recognising the BinderHub and its operators as a supporter of the project.
+
+Names should be short and memorable. They should reflect the name of the entity
+operating the BinderHub. For example a hub operated by "Rubber Duck Inc" should
+use their preferred company name acronym or `rubberduck.mybinder.org`. It
+should not be `woodenhorse.mybinder.org` or something similarly unrelated.
+
+The following paragraphs describe the spirit and "should dos" under which
```
Comment (lheagy) [inline docs/binder/subdomains.rst]:
```suggestion
The following paragraphs describe the spirit and "should do's" under which
```

File: docs/binder/subdomains.rst [hunk 3]
```diff
@@ -0,0 +1,61 @@
+.. _binder-subdomains:
+
+Subdomains on mybinder.org
+==========================
+
+The following sections describe how the project allocates sub-domains of
```
Comment (lheagy) [inline docs/binder/subdomains.rst]:
```suggestion
The following section describes how the project allocates sub-domains of
```

(right now there is only one section in this doc.)

File: docs/binder/subdomains.rst [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+project wants to recognise those who publicly support the project in order
+to advance our mission.
+
+If the BinderHub stops being functional, or is deemed to not behave in an
+acceptable way, the subdomain may be reclaimed at any time, redirecting back
+to the root mybinder.org.
+
+If you would like to obtain a sub-domain that points at your public BinderHub
+please `open a new issue on our repository <https://github.com/jupyterhub/team-compass/issues/new>`_
+mentioning who you are, what you do and where we can find your BinderHub.
```
Comment (lheagy) [inline docs/binder/subdomains.rst]: Do we want to end with a positive shout-out? e.g. something along the lines of:
```
We look forward to new partnerships and are glad that you are interested in helping advance the mybinder.org mission!
```

**Comment by choldgraf** [comment]
btw ping @sylvaincorlay who asked about this yesterday

**Comment by betatim** [comment]
> This would not only help to prevent fracturing of the ecosystem but also enable things like load balancing (e.g. depending on the usage/health of the different instances) in the future.
However, for that to be possible I think it would be necessary for the new instances (sub domain users) not to be easily banned.

Nods. I think having one set of badge URLs would be good or at least have them so that even when the hub stops existing they don't "rot".

Can you explain a bit how it makes it harder in your scenario if there is some external place that has the ultimate control on how much traffic gets sent from the central "distributer"?

In my mind this only works if we can establish trust amongst those in the rotation. This means we can set the rules together and also formulate them so that it makes it easier to obtain buy in from the bosses and funders (with obvious constraints from an operational side and wanting to provide a high quality service).

---

You could imagine that users set a cookie that says "no matter what send me to hub X" (similar to what Matthias proposed and what https://bndr.it does).

---

My thinking on load balancing go in the direction of trying to automate it as much as possible. Including checking for versions on the hubs in the rotation as part of the health check (that is the motivation for https://github.com/jupyterhub/binderhub/pull/821).

**Comment by arnim** [comment]
@betatim I see it pretty much the same way. In the end it will depend on whether the funders are in.

On the technical side, a student of us is currently experimenting with something [similar](https://github.com/gesiscss/binder_gallery/commit/232ebe40f6889ebd2a6265395ef8ca2ac4c5da38) to bndr.it for our gallery to allow also mybinder or pangeo to be used.

File: docs/binder/subdomains.rst (hunk 4 above)
Comment (betatim) [reply docs/binder/subdomains.rst]: Added a slight variation of this.

**Comment by betatim** [comment]
I added a sentence as suggested by @lheagy. Time to merge?

**Pull request merged**